### PR TITLE
Numpy116 migration

### PR DESCRIPTION
**Pull request by hmaarrfk**
@conda-forge/arm-arch I don't have right on the bot's branch. so here goes.
Checklist
* [ ] Used a fork of the feedstock to propose changes
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/conda_smithy.html#how-to-re-render ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by hmaarrfk** [comment]
@conda-forge/arm-arch do any of you have time to look at these failures? Feel free to push to my branch, these are kinda beyond my skill level.

**Comment by hmaarrfk** [comment]
@conda-forge-admin please rerender

**Comment by hmaarrfk** [comment]
Executable file is not ELF:

https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=26891&view=logs&jobId=544de720-5bd1-54e2-b0a2-e430dc2dba91&taskId=8befd53e-126a-5450-1c8e-4ab9ea027269&lineStart=3577&lineEnd=3589&colStart=1&colEnd=69

**Comment by hmaarrfk** [comment]
@tylerjereddy might you know what would have changed between numpy 1.15 and 1.16 that could cause that obscure error?

We can get numpy 1.15 to compile on our systems.

**Comment by mariusvniekerk** [comment]
This may require using shippable (or other native alternatives) for aarch64

**Comment by hmaarrfk** [comment]
did you have shippable rerendering working?

File: shippable.yml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+
+build:
+  pre_ci_boot:
+    # It really doesn't like the centos image.....
+    # image_name: arm64v8/centos
+    # tag_name: 7
+
+    image_name: arm64v8/ubuntu
+    image_tag: 18.04
```
Comment (isuruf) [inline shippable.yml]: Building this using ubuntu 18.04 would make the package built here incompatible with other centos7 and all the arm64v8 builds using numpy will have to be compiled with ubuntu 18.04 or newer.

Any reason not to use the `run_docker_build.sh` script?

Comment (mariusvniekerk) [reply shippable.yml]: mostly just testing for now, to see if the non-native build is the cause of the compilation issue.  Appears that it isn't the case

Comment (hmaarrfk) [reply shippable.yml]: The exact container here is beside the point. I'm just hacking here on some old build scripts I have.

Shippable needs things like sudo, and very particular containers. The original error is still there, so I've ruled out the hypothesis that it was due to qemu

```
    numpy/core/src/multiarray/descriptor.c:2053:9: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
             Py_RETURN_FALSE;
             ^~~~~~~~~~~~~~~
    In file included from numpy/core/include/numpy/npy_math.h:548:0,
                     from numpy/core/src/npymath/npy_math_common.h:9,
                     from numpy/core/src/npymath/npy_math_complex.c.src:34:
    numpy/core/src/npymath/npy_math_internal.h.src: In function 'npy_cacoshf':
    numpy/core/src/npymath/npy_math_internal.h.src:482:12: internal compiler error: Segmentation fault
         return @kind@@c@(x, y);
                ^~~~~~~~~~~~~~~
    Please submit a full bug report,
    with preprocessed source if appropriate.
```

**Comment by hmaarrfk** [comment]
@mariusvniekerk the build error is still there, even in shippable
```
    numpy/core/src/multiarray/descriptor.c:2053:9: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
             Py_RETURN_FALSE;
             ^~~~~~~~~~~~~~~
    In file included from numpy/core/include/numpy/npy_math.h:548:0,
                     from numpy/core/src/npymath/npy_math_common.h:9,
                     from numpy/core/src/npymath/npy_math_complex.c.src:34:
    numpy/core/src/npymath/npy_math_internal.h.src: In function 'npy_cacoshf':
    numpy/core/src/npymath/npy_math_internal.h.src:482:12: internal compiler error: Segmentation fault
         return @kind@@c@(x, y);
                ^~~~~~~~~~~~~~~
    Please submit a full bug report,
    with preprocessed source if appropriate.
```

**Comment by isuruf** [comment]
@hmaarrfk, try unsetting `CFLAGS` to see if some of the default CFLAGS is causing the ICE.

**Comment by hmaarrfk** [comment]
@isuruf do you think it is tree vectorize?
```
CFLAGS=-ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O3 -pipe -I$PREFIX/include -fdebug-prefix-map=${SRC_DIR}=/usr/local/src/conda/${PKG_NAME}-${PKG_VERSION} -fdebug-prefix-map=${PREFIX}=/usr/local/src/conda-prefix
```

File: shippable.yml [hunk 2]
```diff
@@ -0,0 +1,71 @@
+language: none
+
+runtime:
+    nodePool: shippable_shared_aarch64
+    # nodePool: shippable_shared_aarch32
+
+env:
+  global:
+  # The BINSTAR_TOKEN secure variable. This is defined canonically in conda-forge.yml.
+    - secure: bkTdATvev7sVFsP62xFV2ck215nXEtH7eWXdhzRRtbzeKquSkNhTGTCoa5FcLDvAVe36w+Sv59/3/oWNyMood8pIWjHLMC5CqqLdc4NRmyyaCKWys4CLhTTurIBPFSWUilxZW1KCKv/WHOe+zQDi2o9R9lf5/MizuwThHSQOIcqeTIn4wtPzbne5MeKSW+mRCsb+l4E/Q1oY2w/mTJ+izDWkxefstZ2t8RqOxH6H20wwNOOj/1WdeztdCOtCAl99r8Aj58odGyfUMAEyw89c5HglAEPurBQs21DZbHp10NmgSLyIbukplulRUm+cQ37loT/hFfTjPUCqLEC3lu6SPw==
```
Comment (hmaarrfk) [inline shippable.yml]: this binstar token is wrong too.

File: recipe/build.sh [hunk 3]
```diff
@@ -1,10 +1,17 @@
...
 cat > site.cfg <<EOF
 [DEFAULT]
 libraries = blas,cblas,lapack
 library_dirs = $PREFIX/lib
 include_dirs = $PREFIX/include
 EOF
 
+# CFLAGS=-ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O3 -pipe -I$PREFIX/include -fdebug-prefix-map=${SRC_DIR}=/usr/local/src/conda/${PKG_NAME}-${PKG_VERSION} -fdebug-prefix-map=${PREFIX}=/usr/local/src/conda-prefix
+CFLAGS="$CFLAGS -O2"
+CPPFLAGS="$CPPFLAGS -O2"
```
Comment (isuruf) [inline recipe/build.sh]: Can you do this only for aarch64? And also try going back to azure.

**Comment by hmaarrfk** [comment]
@conda-forge-admin, please rerender

File: recipe/build.sh (hunk 3 above)
Comment (hmaarrfk) [reply recipe/build.sh]: mostly done. gotta see what MACHINE is for aarch64

**Comment by hmaarrfk** [comment]
better, thanks! Why do you like to remove `set -ex`. it is so useful for `sh` noobs like me.

**Comment by isuruf** [comment]
`set -e` is the default when conda-build invokes the script. `set -x` doesn't do anything other than echoing the command

**Comment by hmaarrfk** [comment]
i like to echo ;). it teaches me when I need to add quotes.

**Comment by hmaarrfk** [comment]
I'll delete shippable stuff when the build passes.

**Comment by isuruf** [comment]
> i like to echo ;). it teaches me when I need to add quotes.

Sure. Please add it for other branches as well. I'd like to keep the build script same across branches.

**Comment by hmaarrfk** [comment]
@isuruf seems like it could be an issue with char *

https://github.com/numpy/numpy/issues/12862

I mean for now, you already reverted the `set -x` so I won't take the effort for something that works.

**Comment by isuruf** [comment]
@conda-forge-admin, rerender

**Comment by hmaarrfk** [comment]
what changed?

**Comment by isuruf** [comment]
> what changed?

`target_platform` was used and needs a rerender for that variable to be registered as used. I'm not sure why conda-build's default value was not used

File: recipe/build.sh [hunk 4]
```diff
@@ -10,7 +10,7 @@ include_dirs = $PREFIX/include
 EOF
 
 # CFLAGS=-ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O3 -pipe -I$PREFIX/include -fdebug-prefix-map=${SRC_DIR}=/usr/local/src/conda/${PKG_NAME}-${PKG_VERSION} -fdebug-prefix-map=${PREFIX}=/usr/local/src/conda-prefix
-CFLAGS="-I$PREFIX/include -fdebug-prefix-map=${SRC_DIR}=/usr/local/src/conda/${PKG_NAME}-${PKG_VERSION} -fdebug-prefix-map=${PREFIX}=/usr/local/src/conda-prefix -O2"
+CFLAGS="$CFLAGS -O2"
```
Comment (isuruf) [inline recipe/build.sh]: `-ftree-vectorize` might still be an issue

**Comment by tylerjereddy** [comment]
@hmaarrfk Not sure what's going on here from a very quick look--is there something specific  you want from the NumPy side of things? We have Shippable ARMv8 runs in our CI & I've even built OpenBLAS on ARMv8 to hopefully better match wheels on other platforms in the future.

Just scanning through--are there specific flags we should test for / constrain on, that are needed for ARM sometimes but not currently used in our CI?

**Comment by hmaarrfk** [comment]
@tylerjereddy I guess I'm wondering if you all changed something that would have an effect on the necessary compiler flags in 1.16. If not, then it is probably on our side.

**Comment by hmaarrfk** [comment]
if somebody wants to cancel all the previous shippable builds that would be good for when we need them

**Comment by tylerjereddy** [comment]
There were some tricky changes in the way that distutils handles compiler flags, although many of the problems that arose were reported & patched, I think.

See, for example, the 80-comment PR on `exec_command` removal: https://github.com/numpy/numpy/pull/11898

I'm not sure that is the problem. I also see @rgommers talking about some recent conda-related matters in the last 24 hours here: https://github.com/numpy/numpy/issues/13280

But again, not sure that is related. I just looked at one of the failed aarch64 builds here, and there is just one precision-related test failure. Those Azure runs aren't running on native ARMv8 I presume, so perhaps looking less serious now than before with just that failure?

**Comment by rgommers** [comment]
I check a couple of runs, they all have only one failure: `test_loss_of_precision[complex256]`. That's definitely not related to any of the `numpy.distutils` changes in 1.16, and very likely just needs a small test precision bump. It's hard to see because `assert_(np.all(d < rtol)` isn't particularly informative (should have been `assert_allclose`).

**Comment by hmaarrfk** [comment]
Sorry @rgommers, the most recent run use `-O0` which disables optimizations. I think we would consider this a very undesirable fallback.

This build uses the default `-O3` flags 
https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=26908

and fails with
```
    numpy/core/src/multiarray/descriptor.c:2053:9: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
             Py_RETURN_FALSE;
             ^~~~~~~~~~~~~~~
    In file included from numpy/core/include/numpy/npy_math.h:548:0,
                     from numpy/core/src/npymath/npy_math_common.h:9,
                     from numpy/core/src/npymath/npy_math_complex.c.src:34:
    numpy/core/src/npymath/npy_math_internal.h.src: In function 'npy_cacoshf':
    numpy/core/src/npymath/npy_math_internal.h.src:482:12: internal compiler error: Segmentation fault
         return @kind@@c@(x, y);
                ^~~~~~~~~~~~~~~
    Please submit a full bug report,
    with preprocessed source if appropriate.
```

**Comment by rgommers** [comment]
Hmm fun. Okay read the whole issue now - seems tricky, doesn't ring any bells. Sorry for the noise.

**Comment by hmaarrfk** [comment]
Thanks for chiming in

**Comment by jjhelmus** [comment]
I was able to reproduce the internal compiler error on real aarch64 hardware using the conda provided compilers as well as gcc 7.3.0 from Ubuntu 18.04 using the preprocessed source generated in the docker image.  The preprocessed source that is giving the error can be found at http://ix.io/1FOb.  Compiling with an optimization level of `-O0` or without `-fPIC` will succeed but otherwise the compilation fails with a segmentation fault:

```
$ gcc -O1 -fPIC -c npy_math_complex.prepoc.c 
In file included from numpy/core/include/numpy/npy_math.h:548:0,
                 from numpy/core/src/npymath/npy_math_common.h:9,
                 from numpy/core/src/npymath/npy_math_complex.c:42:
numpy/core/src/npymath/npy_math_internal.h.src: In function 'npy_cacoshf':
numpy/core/src/npymath/npy_math_internal.h.src:482:12: internal compiler error: Segmentation fault
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-7/README.Bugs> for instructions.
```

**Comment by jjhelmus** [comment]
And here is a minimum working example of the ICE:

``` C
#include <math.h>

typedef struct { float real, imag; } cfloat;

static inline float cimag_cfloat(cfloat z)
{
    union { cfloat z; float a[2]; } __z_repr;
    __z_repr.z = z; 
    return __z_repr.a[1];
}

float
bug(cfloat z)
{
    float f1, f2;
    f1 = cimag_cfloat(z);
    f2 = copysignf(0, f1);
    return f2;
}
```

```
$ gcc --version
gcc (Ubuntu/Linaro 7.3.0-27ubuntu1~18.04) 7.3.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
$ gcc -O1 -fPIC -c mre.c 
mre.c: In function 'bug':
mre.c:17:8: internal compiler error: Segmentation fault
     f2 = copysignf(0, f1);
     ~~~^~~~~~~~~~~~~~~~~~
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-7/README.Bugs> for instructions.
```
I'm curious if this was fixed in gcc 7.4 or 8.

**Comment by hmaarrfk** [comment]
That minimum reproducible example is awesome.

**Comment by jjhelmus** [comment]
Unfortunately the aarch64 ICE is not fixed in gcc 7.4 or 8.

**Comment by isuruf** [comment]
How about clang? clangdev 8.0.0 on conda-forge now works with the anaconda gcc sysroot. We can build it for aarch64 outside of CI.

**Comment by hmaarrfk** [comment]
can we patch numpy around this bug?

**Comment by jjhelmus** [comment]
> How about clang?

I do not know of an easy method of testing clang on aarch64.  GCC provides [docker images](https://hub.docker.com/_/gcc/) for the arm64v8 platform that makes testing easy.  I could not find anything similar for clang.

**Comment by isuruf** [comment]
I was suggesting using the conda-forge recipe, but that might take a long time to build. Btw, I tried gcc4 that is installed when you do `yum install gcc` on condaforge/linux-anvil-aarch64 and I couldn't reproduce the error.

**Comment by isuruf** [comment]
Here's an easy way to check clang. Download http://releases.llvm.org/8.0.0/clang+llvm-8.0.0-aarch64-linux-gnu.tar.xz and a recent libstdcxx from conda and cp libstdc++ to the lib folder.

**Comment by isuruf** [comment]
I can reproduce on docker. I also used clang-8 and it doesn't have this issue.

**Comment by mariusvniekerk** [comment]
@isuruf, we can always build that offline somewhere using that recipe,  I have a cloud aarch64 instance exactly for that purpose

**Comment by isuruf** [comment]
> @isuruf, we can always build that offline somewhere using that recipe, I have a cloud aarch64 instance exactly for that purpose

Can you build llvmdev and clangdev then? All the dependencies should be available on conda-forge and c4aarch64.

> if you have substantial mono experience i'd also accept it for building the azure build-agent :D

C#? Nope. :smile:

**Comment by mariusvniekerk** [comment]
Okay, i've built clangdev 8 for aarch64.

**Comment by isuruf** [comment]
@mariusvniekerk, need to rebuild clangdev with the latest changes

**Comment by mariusvniekerk** [comment]
Looks like its building properly now

**Comment by mariusvniekerk** [comment]
Seems we need the patch to skip the longcomplex tests here too

**Comment by mariusvniekerk** [comment]
And probably build ppc64le on native

**Comment by hmaarrfk** [comment]
For ppc64le did you have anything else to do before pushing your travis scripts upstream marius?

**Comment by mariusvniekerk** [comment]
@conda-forge-admin please rerender

**Comment by mariusvniekerk** [comment]
@conda-forge-admin please rerender

**Comment by hmaarrfk** [comment]
Can somebody with knowledge of PPC64LE comment on whether or not this is a real error, or a bug with the tests?

```python
___________ TestComplexFunctions.test_loss_of_precision[complex256] ____________

self = <numpy.core.tests.test_umath.TestComplexFunctions object at 0x3fff78526310>
dtype = <type 'numpy.complex256'>

    @pytest.mark.parametrize('dtype', [np.complex64, np.complex_, np.longcomplex])
    def test_loss_of_precision(self, dtype):
        """Check loss of precision in complex arc* functions"""
    
        # Check against known-good functions
    
        info = np.finfo(dtype)
        real_dtype = dtype(0.).real.dtype
        eps = info.eps
    
        def check(x, rtol):
            x = x.astype(real_dtype)
    
            z = x.astype(dtype)
            d = np.absolute(np.arcsinh(x)/np.arcsinh(z).real - 1)
            assert_(np.all(d < rtol), (np.argmax(d), x[np.argmax(d)], d.max(),
                                      'arcsinh'))
    
            z = (1j*x).astype(dtype)
            d = np.absolute(np.arcsinh(x)/np.arcsin(z).imag - 1)
            assert_(np.all(d < rtol), (np.argmax(d), x[np.argmax(d)], d.max(),
                                      'arcsin'))
    
            z = x.astype(dtype)
            d = np.absolute(np.arctanh(x)/np.arctanh(z).real - 1)
            assert_(np.all(d < rtol), (np.argmax(d), x[np.argmax(d)], d.max(),
                                      'arctanh'))
    
            z = (1j*x).astype(dtype)
            d = np.absolute(np.arctanh(x)/np.arctan(z).imag - 1)
            assert_(np.all(d < rtol), (np.argmax(d), x[np.argmax(d)], d.max(),
                                      'arctan'))
    
        # The switchover was chosen as 1e-3; hence there can be up to
        # ~eps/1e-3 of relative cancellation error before it
    
        x_series = np.logspace(-20, -3.001, 200)
        x_basic = np.logspace(-2.999, 0, 10, endpoint=False)
    
        if dtype is np.longcomplex:
            # It's not guaranteed that the system-provided arc functions
            # are accurate down to a few epsilons. (Eg. on Linux 64-bit)
            # So, give more leeway for long complex tests here:
            # Can use 2.1 for > Ubuntu LTS Trusty (2014), glibc = 2.19.
>           check(x_series, 50.0*eps)

../_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib/python2.7/site-packages/numpy/core/tests/test_umath.py:2596: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

x = array([1.00000000e-20, 1.21736864e-20, 1.48198641e-20, 1.80412378e-20,
       ...871e-04, 6.73218092e-04, 8.19554595e-04, 9.97700064e-04],
      dtype=float128)
rtol = 1.23259516440783094595582588325435e-30

    def check(x, rtol):
        x = x.astype(real_dtype)
    
        z = x.astype(dtype)
        d = np.absolute(np.arcsinh(x)/np.arcsinh(z).real - 1)
        assert_(np.all(d < rtol), (np.argmax(d), x[np.argmax(d)], d.max(),
>                                 'arcsinh'))
E       AssertionError: (131, 1.54987431981169249551435343964035e-09, 4.0035173453529620614007210953362e-19, 'arcsinh')
```

**Comment by hmaarrfk** [comment]
Well if somebody can confirm that the ppc64le are OK failures, I think I just included the patch in the recipe, and the selector would just have to be changed.

**Comment by tylerjereddy** [comment]
@hmaarrfk Are you using genuine ppc64le architecture? I don't see shippable in the CI matrix for this PR & we don't see the failure on NumPy master with ppc64le / shippable.

That said, I **do** see the exact failure you mention when I run the NumPy test suite on WSL (Windows Subsystem for Linux), so I wouldn't be surprised if some other kind of "emulation" experiences a similar failure.

I'm sure floating point calc experts know far more about this stuff..

**Comment by hmaarrfk** [comment]
@tylerjereddy I think marius got native PPC64LE running on travis
https://travis-ci.org/conda-forge/numpy-feedstock/builds/525758989

Any idea why windows is randomly failing?

**Comment by tylerjereddy** [comment]
oh, right I'm confusing ppc64le and ARM -- don't think we see that on master either though, since we do the Travis ppc64le testing too

**Comment by hmaarrfk** [comment]
yeah, thats why im half worried. ill try to see if any prs were made affecting that in master.


Do you mind backporting 
https://github.com/numpy/numpy/pull/12709
Just to see the results on 1.16.3

**Comment by hmaarrfk** [comment]
This seems to be the  most recent backport that touches that part of the code.
https://github.com/numpy/numpy/pull/8566
https://github.com/numpy/numpy/pull/12378

**Comment by hmaarrfk** [comment]
@conda-forge-admin please rerender

**Comment by hmaarrfk** [comment]
<details> <summary> Is this OK??? </summary>

```python
self = <numpy.core.tests.test_einsum.TestEinsum object at 0x3fff8be3c8d0>

    def test_einsum_sums_cfloat64(self):
>       self.check_einsum_sums('c8')

../_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib/python3.6/site-packages/numpy/core/tests/test_einsum.py:561: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <numpy.core.tests.test_einsum.TestEinsum object at 0x3fff8be3c8d0>
dtype = 'c8', do_opt = False

    def check_einsum_sums(self, dtype, do_opt=False):
        # Check various sums.  Does many sizes to exercise unrolled loops.
    
        # sum(a, axis=-1)
        for n in range(1, 17):
            a = np.arange(n, dtype=dtype)
            assert_equal(np.einsum("i->", a, optimize=do_opt),
                         np.sum(a, axis=-1).astype(dtype))
            assert_equal(np.einsum(a, [0], [], optimize=do_opt),
                         np.sum(a, axis=-1).astype(dtype))
    
        for n in range(1, 17):
            a = np.arange(2*3*n, dtype=dtype).reshape(2, 3, n)
            assert_equal(np.einsum("...i->...", a, optimize=do_opt),
                         np.sum(a, axis=-1).astype(dtype))
            assert_equal(np.einsum(a, [Ellipsis, 0], [Ellipsis], optimize=do_opt),
                         np.sum(a, axis=-1).astype(dtype))
    
        # sum(a, axis=0)
        for n in range(1, 17):
            a = np.arange(2*n, dtype=dtype).reshape(2, n)
            assert_equal(np.einsum("i...->...", a, optimize=do_opt),
                         np.sum(a, axis=0).astype(dtype))
            assert_equal(np.einsum(a, [0, Ellipsis], [Ellipsis], optimize=do_opt),
                         np.sum(a, axis=0).astype(dtype))
    
        for n in range(1, 17):
            a = np.arange(2*3*n, dtype=dtype).reshape(2, 3, n)
            assert_equal(np.einsum("i...->...", a, optimize=do_opt),
                         np.sum(a, axis=0).astype(dtype))
            assert_equal(np.einsum(a, [0, Ellipsis], [Ellipsis], optimize=do_opt),
                         np.sum(a, axis=0).astype(dtype))
    
        # trace(a)
        for n in range(1, 17):
            a = np.arange(n*n, dtype=dtype).reshape(n, n)
            assert_equal(np.einsum("ii", a, optimize=do_opt),
                         np.trace(a).astype(dtype))
            assert_equal(np.einsum(a, [0, 0], optimize=do_opt),
                         np.trace(a).astype(dtype))
    
        # multiply(a, b)
        assert_equal(np.einsum("..., ...", 3, 4), 12)  # scalar case
        for n in range(1, 17):
            a = np.arange(3 * n, dtype=dtype).reshape(3, n)
            b = np.arange(2 * 3 * n, dtype=dtype).reshape(2, 3, n)
            assert_equal(np.einsum("..., ...", a, b, optimize=do_opt),
                         np.multiply(a, b))
            assert_equal(np.einsum(a, [Ellipsis], b, [Ellipsis], optimize=do_opt),
                         np.multiply(a, b))
    
        # inner(a,b)
        for n in range(1, 17):
            a = np.arange(2 * 3 * n, dtype=dtype).reshape(2, 3, n)
            b = np.arange(n, dtype=dtype)
>           assert_equal(np.einsum("...i, ...i", a, b, optimize=do_opt), np.inner(a, b))
E           AssertionError: 
E           Arrays are not equal
E           
E           Mismatch: 50%
E           Max absolute difference: 28.
E           Max relative difference: 0.056
E            x: array([[ 204.+0.j,  528.+0.j,  852.+0.j],
E                  [1176.+0.j, 1500.+0.j, 1824.+0.j]], dtype=complex64)
E            y: array([[ 204.+0.j,  500.+0.j,  852.+0.j],
E                  [1148.+0.j, 1500.+0.j, 1796.+0.j]], dtype=complex64)

```
</details>

**Comment by hmaarrfk** [comment]
@conda-forge-admin, please rerender

**Comment by hmaarrfk** [comment]
sigh the aarch build got stuck, but i'll save CI resources and just leave it at that.

**Comment by hmaarrfk** [comment]
@conda-forge/arm-arch does anybody want to give this a last review and a good push?

**Comment by hmaarrfk** [comment]
I guess we just update the pinning in recipes to be numpy 1.16? and call it a day?

**Pull request merged**